packages/x-data-grid/src/components/GridSkeletonLoadingOverlay.tsx

```diff
@@ -0,0 +1,135 @@
+  HTMLDivElement,
+  React.HTMLAttributes<HTMLDivElement>
+>(function GridSkeletonLoadingOverlay(props, forwardedRef) {
+  const ref = React.useRef<HTMLDivElement>(null);
+  const handleRef = useForkRef(ref, forwardedRef);
+
+  const apiRef = useGridApiContext();
+
+  const dimensions = apiRef.current?.getRootDimensions();
+  const viewportHeight = dimensions?.viewportInnerSize.height ?? 0;
+
+  // @ts-expect-error Function signature expects to be called with parameters, but the implementation suggests otherwise
+  const rowHeight = apiRef.current.unstable_getRowHeight();
+  const skeletonRowsCount = Math.ceil(viewportHeight / rowHeight);
+
+  const totalWidth = gridColumnsTotalWidthSelector(apiRef);
+  const positions = gridColumnPositionsSelector(apiRef);
+  const inViewportCount = React.useMemo(
+    () => positions.filter((value) => value <= totalWidth).length,
+    [totalWidth, positions],
+  );
+  const columns = apiRef.current.getVisibleColumns().slice(0, inViewportCount);
+
+  const children = React.useMemo(() => {
+    // reseed random number generator to create stable lines betwen renders
+    const array: React.ReactNode[] = [];
+    const randomNumberBetween = seededRandomNumberGenerator(12345);
+
+    for (let i = 0; i < skeletonRowsCount; i += 1) {
+      // eslint-disable-next-line no-restricted-syntax
+      for (const column of columns) {
+        const [min, max] = column.type
+          ? COLUMN_WIDTH_RANGE_BY_TYPE[column.type] ?? DEFAULT_COLUMN_WIDTH_RANGE
+          : DEFAULT_COLUMN_WIDTH_RANGE;
+        const width = Math.round(randomNumberBetween(min, max));
+        const isCircular = column.type === 'boolean' || column.type === 'actions';
+
+        array.push(
+          <SkeletonCell
+            key={`skeleton-column-${i}-${column.field}`}
+            sx={{ justifyContent: column.align }}
+          >
+            <Skeleton
+              width={isCircular ? '1.3em' : `${width}%`}
+              height={isCircular ? '1.3em' : '1.2em'}
+              variant={isCircular ? 'circular' : 'text'}
+              sx={{ mx: 1 }}
+            />
+          </SkeletonCell>,
+        );
+      }
+      array.push(<SkeletonCell key={`skeleton-filler-column-${i}`} />);
+    }
+    return array;
+  }, [skeletonRowsCount, columns]);
+
+  const [initialColWidthVariables, gridTemplateColumns] = columns.reduce(
+    ([initialSize, templateColumn], column, i) => {
+      const varName = colWidthVar(i);
+      initialSize[varName] = `${column.computedWidth}px`;
+      templateColumn += ` var(${varName})`;
+      return [initialSize, templateColumn];
+    },
+    [{} as Record<string, string>, ''],
+  );
+
+  React.useEffect(() => {
+    const handleScrollChange: GridEventListener<'scrollPositionChange'> = (params) => {
+      if (ref.current) {
+        ref.current.scrollLeft = params.left;
+      }
+    };
+    return apiRef.current.subscribeEvent('scrollPositionChange', handleScrollChange);
+  }, [apiRef]);
```

Comment:
We have a hook that does this, I think `useGridEventHandler('scrollPositionChange', (...) => {...})`. I think the name is wrong but it's something like that.

Reply:
Fixed https://github.com/mui/mui-x/pull/13293/commits/d45dfddcd90f0b572d548992a8ae2e6a22884882